src/combinatorics/rado/rado_linear_algebra.lean

```diff
@@ -0,0 +1,135 @@
+`#s ≤ rank (span (f '' s))` for all finite subsets `s` of `ι`; see
+`linear_independent.iff_card_le_rank_span_on_finsets`.
+
+We then use this show the linear algebra version of Rado's Theorem:
+
+*`F : ι → finset M` has the property that `#s ≤ rank (span ⋃ i in s, F i)`
+for all finite subsets `s` of `\io` if and only if there exists a linearly independent
+family `f : ι → M` such that `f i ∈ F i` for all `i` ("`f` is a section of `F`").*
+
+See `linear_independent.rado`.
+-/
+
+namespace submodule
+
+/-- The rank of a submodule spanned by a single element is at most one. -/
+lemma finrank_span_singleton_le {K V : Type*} [division_ring K] [add_comm_group V] [module K V]
+  (v : V) : finite_dimensional.finrank K (span K ({v} : set V)) ≤ 1 :=
+begin
+  rcases eq_or_ne v 0 with h | h,
+  { rw [span_singleton_eq_bot.mpr h, finrank_bot],
+    exact zero_le_one, },
+  { rw finrank_span_singleton h, }
+end
+
+end submodule
+
+namespace linear_independent
+
+open finite_dimensional submodule rank_fn finset
+
+variables {R : Type*} [division_ring R]
+          {M : Type*} [add_comm_group M] [decidable_eq M] [module R M]
+
+/-- A convenience lemma: if `s` and `insert a s` span the same submodule, then
+any submodule containing `s` also contains `a` (version for a finset `s`). -/
+lemma mem_submodule_of_rank_eq_of_finset_subset {s : finset M} {a : M}
+  (h: finrank R (span R (↑(has_insert.insert a s) : set M)) = finrank R (span R (s : set M)))
+  (p : submodule R M) (hsp : (s : set M) ⊆ p) : a ∈ p :=
+begin
+  have ha : a ∈ span R (↑(has_insert.insert a s) : set M),
+  { rw [coe_insert],
+    exact subset_span (set.mem_insert a ↑s), },
+  rw ← eq_of_le_of_finrank_eq (span_mono $ coe_subset.mpr (subset_insert _ _)) h.symm at ha,
+  exact mem_span.mp ha p hsp,
+end
+
+/-- If `rank (span ({a} ∪ s)) = rank (span s) = rank (span ({b} ∪ s))`,
+then `rank (span ({a, b} ∪ s)) = rank (span s)` -/
+lemma submodule_insert_insert_eq_of_submodule_insert_eq {s : finset M} {a b : M}
+  (ha: finrank R (span R (↑(has_insert.insert a s) : set M)) = finrank R (span R (s : set M)))
+  (hb: finrank R (span R (↑(has_insert.insert b s) : set M)) = finrank R (span R (s : set M))) :
+  span R (↑(has_insert.insert b (has_insert.insert a s)) : set M) = span R (s : set M) :=
+begin
+  ext m,
+  simp_rw mem_span,
+  refine ⟨λ h' (p : submodule R M) hi, _, λ h' (p : submodule R M) hi, _⟩,
+  { refine h' p _,
+    simp_rw [coe_insert],
+    exact set.insert_subset.mpr ⟨mem_submodule_of_rank_eq_of_finset_subset hb p hi,
+            set.insert_subset.mpr ⟨mem_submodule_of_rank_eq_of_finset_subset ha p hi, hi⟩⟩, },
+  { rw [coe_insert, coe_insert] at hi,
+    exact h' p (((set.subset_insert a ↑s).trans (set.subset_insert b _)).trans hi), }
+end
+
+variables (R M)
+
+/-- The rank of the submodule generated by a finset defines a rank function. -/
+noncomputable
+def rank_rk : rank_fn M :=
+{ to_fun := λ s, finrank R (span R (s : set M)), -- there is `set.finrank`, but no API for it...
```

Comment:
You should use `set.finrank` here anyway; you can always unfold it to get back to this definition.

There is API for it, but it's all misnamed so it's hard to find. See for instance

https://github.com/leanprover-community/mathlib/blob/039ef89bef6e58b32b62898dd48e9d1a4312bb65/src/linear_algebra/finrank.lean#L292-L295

Reply:
OK; I'll have a look what's there and try to use it. Probably searching for ".finrank" in mathlib should help.

Reply:
I changed it, but I still maintain that there is (almost) no API for `set.finrank`. The only relevant lemmas I could find are
* [finrank_span_finset_le_card](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finrank.html#finrank_span_finset_le_card)
* [finrank_range_le_card](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finrank.html#finrank_range_le_card)
* [linear_independent_iff_card_eq_finrank_span](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finrank.html#linear_independent_iff_card_eq_finrank_span)
* [linear_independent_iff_card_le_finrank_span](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finrank.html#linear_independent_iff_card_le_finrank_span)

I would at least expect things like `set.finrank_empty` or `set.finrank_mono`...

Reply:
BTW, I think it would make (more?) sense to have `finset.finrank` -- if you want finite rank, finite generating sets are a natural thing to consider.

Reply:
I just saw that [set.finrank_mono](https://leanprover-community.github.io/mathlib_docs/linear_algebra/finite_dimensional.html#set.finrank_mono) does exist (it sits in a different file, though...). I wonder why I did not find it grepping for `.finrank `.